### PR TITLE
Align notification APIs with new KVO APIs

### PR DESCRIPTION
**Pull request by jpsim**
Addresses #5025. I opted to leave the Objective-C `addNotificationBlock` API as-is, because `observe` isn't a very idiomatic ObjC method name. However, the ObjC API _is_ affected by this (i.e. `-[RLMNotificationToken invalidate]`) in order to keep our terminology aligned between Swift & ObjC and the first argument of idiomatic naming doesn't apply here.

### CODE REVIEW

**Comment by austinzheng** [comment]
It looks like the object server tests may need to be updated; they seem to not have built.

**Comment by jpsim** [comment]
Fixed in 6e96202

**Comment by austinzheng** [comment]
Thanks.

Sorry for making you have to do that; eventually we'll be getting rid of `RLMSyncPermissionsResults` as its own base type altogether (but we still need the changes to get your PR in; it shouldn't be blocked on my stuff).

**Comment by jpsim** [comment]
I'll add the missing docs tomorrow morning.

**Comment by tgoyne** [comment]
What impact does this have on using the native `observe()` function on `Object`? The method signatures are sufficiently different that I wouldn't expect any issues other than slightly worse autocomplete, but we don't have any tests which use it.

**Comment by austinzheng** [comment]
It looks like there are still a bunch of KVO tests failing. Not sure if they're related.

**Comment by jpsim** [comment]
Only failing on Xcode 8.0. I guess I have to download that now.

File: RealmSwift/LinkingObjects.swift [hunk 1]
```diff
@@ -436,10 +436,3 @@ internal enum LinkingObjectsBridgingMetadata {
         }
     }
 }
-
-// MARK: Unavailable
```
Comment (austinzheng) [inline RealmSwift/LinkingObjects.swift]: Why was this removed? It provides Xcode fix-its for users who are migrating from Realm 2.* to 3 and still using the deprecated APIs.

File: RealmSwift/List.swift [hunk 2]
```diff
@@ -524,10 +524,3 @@ extension List: AssistedObjectiveCBridgeable {
         return (objectiveCValue: _rlmArray, metadata: nil)
     }
 }
-
-// MARK: Unavailable
-
-extension List {
```
Comment (austinzheng) [inline RealmSwift/List.swift]: Ditto.

**Comment by austinzheng** [comment]
Looks good, my comments notwithstanding. Looks like a rebase is needed.

File: RealmSwift/LinkingObjects.swift (hunk 1 above)
Comment (jpsim) [inline RealmSwift/LinkingObjects.swift]: Because the migration assistance this provides is already provided by the more generic ones on `RealmCollection`, so it's redundant to add it here as well.

File: RealmSwift/List.swift (hunk 2 above)
Comment (jpsim) [inline RealmSwift/List.swift]: Because the migration assistance this provides is already provided by the more generic ones on `RealmCollection`, so it's redundant to add it here as well.
